api/logs/v1/labels_enforcer_test.go

```diff
@@ -0,0 +1,114 @@
+		desc           string
+		accessMatchers []*labels.Matcher
+		urlValues      url.Values
+		expValues      url.Values
+	}{
+		{
+			desc: "empty url values",
+			accessMatchers: []*labels.Matcher{
+				{
+					Type:  labels.MatchRegexp,
+					Name:  "kubernetes_namespace_name",
+					Value: "log-test-0",
+				},
+			},
+			urlValues: url.Values{},
+			expValues: url.Values{},
+		},
+		{
+			desc: "query with no matcher overlap",
+			accessMatchers: []*labels.Matcher{
+				{
+					Type:  labels.MatchRegexp,
+					Name:  "kubernetes_namespace_name",
+					Value: "ns-name|another-ns-name|openshift-.*",
+				},
+			},
+			urlValues: url.Values{
+				"query": []string{"{kubernetes_container_name=\"logger\", kubernetes_pod_name=\"pod-name\"}"},
+			},
+			expValues: url.Values{
+				"query": []string{"{kubernetes_namespace_name=~\"ns-name|another-ns-name|openshift-.*\", kubernetes_container_name=\"logger\", kubernetes_pod_name=\"pod-name\"}"},
+			},
+		},
+		{
+			desc: "query with accessible namespace",
+			accessMatchers: []*labels.Matcher{
+				{
+					Type:  labels.MatchRegexp,
+					Name:  "kubernetes_namespace_name",
+					Value: "ns-name|another-ns-name|openshift-.*",
+				},
+			},
+			urlValues: url.Values{
+				"query": []string{"{kubernetes_namespace_name=\"ns-name\", kubernetes_container_name=\"logger\", kubernetes_pod_name=\"pod-name\"}"},
+			},
+			expValues: url.Values{
+				"query": []string{"{kubernetes_namespace_name=\"ns-name\", kubernetes_container_name=\"logger\", kubernetes_pod_name=\"pod-name\"}"},
+			},
+		},
+		{
+			desc: "query with a forbidden namespace",
+			accessMatchers: []*labels.Matcher{
+				{
+					Type:  labels.MatchRegexp,
+					Name:  "kubernetes_namespace_name",
+					Value: "ns-name|another-ns-name|openshift-.*",
+				},
+			},
+			urlValues: url.Values{
+				"query": []string{"{kubernetes_namespace_name=\"forbidden-ns|ns-name\", kubernetes_container_name=\"logger\", kubernetes_pod_name=\"pod-name\"}"},
+			},
+			expValues: url.Values{
+				"query": []string{"{kubernetes_namespace_name=~\"ns-name|another-ns-name|openshift-.*\", kubernetes_container_name=\"logger\", kubernetes_pod_name=\"pod-name\"}"},
+			},
+		},
+	}
+
+	for _, tc := range tt {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
+
+			v, err := enforceValues(tc.accessMatchers, tc.urlValues)
+			require.Nil(t, err)
+
+			if len(tc.urlValues.Encode()) == 0 {
+				// Url values are empty, non need to do more checks.
+				require.Len(t, v, 0)
+				return
+			}
+
+			u, err := url.ParseQuery(v)
+			require.Nil(t, err)
+
+			expr, err := logqlv2.ParseExpr(u.Get("query"))
+			require.Nil(t, err)
+
+			expected, err := logqlv2.ParseExpr(tc.expValues.Get("query"))
+			require.Nil(t, err)
+
+			smExpr := expr.(*logqlv2.StreamMatcherExpr)
+			require.NotNil(t, smExpr)
+
+			smExpected := expected.(*logqlv2.StreamMatcherExpr)
+			require.NotNil(t, smExpected)
+
+			assert.ElementsMatch(t, smExpected.Matchers(), smExpr.Matchers())
```

Comment:
How is implementing a sort and compare simpler that just calling the `ElementsMatch` function?

Reply:
Sorry, let me clarify what I mean by this testify method being complex. 

Sure it's less loc/char when we use it, but the `ElementsMatch` function itself seems to do quite a bit that you simply don't need for a test like this. If you look at the function definition, it calls a `diffLists` method which loops through the two lists while maintaining a `visited` slice, and then calls a `ObjectsAreEqual` method for each comparison between two lists which is basically a `reflect.DeepEqual` call. It accounts for all types with heavy `reflect` usage and for duplicate elements. It abstracts and hides away all of this complexity.

In comparison, `testutil.Equals` is just `reflect.DeepEqual` call and we only need small 4 line type def for sorting the lists and comparing. 

And by this being complex, I do not only mean these methods. Switching to `testify` just for one particular use case that can be accomplished easily with existing test package, doesn't make sense to me. I agree with @periklis that having both `testutil` and `testify` together in the same repo seems like a code smell with multiple ways of writing things, and would need some unanimous decision from current contributors/maintainers.

IMO `testify` is overkill for testing, and abstracts away extreme complexity. It generally has [too many different APIs](https://pkg.go.dev/github.com/stretchr/testify@v1.8.2/assert) for relatively simple functionality that can be ensured in the test itself by the user, and makes tests unnecessarily complex by abstracting away what exactly we are asserting on. In comparison, `testutil` has a very clean and minimal [interface](https://pkg.go.dev/github.com/efficientgo/core/testutil) that allows composing tests in a very explicit fashion. Not to mention we use this consistently everywhere in ecosystem.

So if you do plan to switch to this, I think you'd need to have some explicit reasons for doing so, and feedback on why existing `testutil` implementation doesn't work? 🙂

Reply:
The reason for using `testify` is that it's simple and it works, no need to do mental gymnastic to justify it, it just works better :slightly_smiling_face:

Reply:
Also, `testutil` doesn't do many comparison types (like element matching), thus making it objectively inferior and requiring us to do a lot of the implementation ourselves instead of focusing on what matters, which is the actual test cases.
I understand we can do it all with `testutil` with some additional code, but that's not a good reason. By that logic, why even use `testutil`, let's just implement everything ourselves.
I don't know why you are hung up on this.

Reply:
IMHO this thread discussion goes beyond the scope of this pull request and belongs into the dedicate issue #468. I All involved parties including myself have pro/cons arguments and we should reflect this on the mentioned issue to make the discussion visible to the whole maintainers and contributors group. 

For a circuit breaker, I opt in for the `sortableMatcher` plus `testutil.Equals` proposal to push forward with this PR solving real user issues. In addition both of you can re-assemble your arguments on #468 and we can have a proper discussion with other maintainers/contributors over there.